Use a lock on cleanup, expire_tool and cache_tool

mvdbeek opened the pull request:
This should fix https://github.com/galaxyproject/galaxy/issues/7444.

nsoranzo committed:
Add debugging statements to ``ToolCache.cleanup()``

mvdbeek committed:
Use a lock on cleanup, expire_tool and cache_tool

This should fix https://github.com/galaxyproject/galaxy/issues/7444.

nsoranzo commented:
Damn, at the third re-run it failed.

nsoranzo commented on lib/galaxy/tools/cache.py:
```diff
@@ -79,31 +84,33 @@ def get_tool_by_id(self, tool_id):
+                    self._macro_paths_by_id[tool_id].add(macro_path)
 
     def reset_status(self):
         """
```
Not relevant for the unit test failure, but probably also this method should use the lock.

mvdbeek replied:
Yep, was just looking at that too.

nsoranzo committed:
Use lock also for ``ToolCache.reset_status()`` method

nsoranzo committed:
Use ``_try_until_no_errors`` also for ``test_tool_reload_when_macro_is_altered`` unit test

mvdbeek commented on lib/galaxy/tools/cache.py:
```diff
@@ -69,49 +80,54 @@ def _should_cleanup(self, config_filename):
     def get_tool(self, config_filename):
         """Get the tool at `config_filename` from the cache if the tool is up to date."""
-        return self._tools_by_path.get(config_filename, None)
+        with self._lock:
```
I don't think this is necessary or a good idea, getting things from a dict is an atomic operation. I think we only need to lock add and remove (and I think that's why cleanup was already wrapped in try/except).

mvdbeek committed:
Use ``with`` statement

mvdbeek commented:
So the circle ci log tells me that the guts of the logic is working fine, even if the assertion fails:
```
self = <unit.tools.test_toolbox.ToolBoxTestCase testMethod=test_tool_reload_for_broken_tool>

    def test_tool_reload_for_broken_tool(self):
        self._init_tool(filename="simple_tool.xml", version="1.0")
        self._add_config("""<toolbox><tool file="simple_tool.xml"/></toolbox>""")
        toolbox = self.toolbox
        tool = toolbox.get_tool('test_tool')
        assert tool is not None
        assert tool.version == "1.0"
        assert tool.tool_errors is None
        # Tool is loaded, now let's break it
        tool_path = tool.config_file
        with open(tool.config_file, 'w') as out:
            out.write('certainly not a valid tool')
    
        def check_tool_errors():
            tool = self.app.toolbox.get_tool("test_tool")
            assert tool is not None
            assert tool.version == "1.0"
            assert tool.tool_errors == 'Current on-disk tool is not valid'
    
        self._try_until_no_errors(check_tool_errors)
    
        # Tool is still loaded, lets restore it with a new version
        self._init_tool(filename="simple_tool.xml", version="2.0")
    
        def check_no_tool_errors():
            tool = self.app.toolbox.get_tool("test_tool")
            assert tool is not None
            assert tool.version == "2.0"
            assert tool.tool_errors is None
            assert tool_path == tool.config_file, "config file"
    
>       self._try_until_no_errors(check_no_tool_errors)

test/unit/tools/test_toolbox.py:236: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <unit.tools.test_toolbox.ToolBoxTestCase testMethod=test_tool_reload_for_broken_tool>
f = <function check_no_tool_errors at 0x7fb792404140>

    def _try_until_no_errors(self, f):
        e = None
        for i in range(2):
            try:
                f()
                return
            except AssertionError as error:
                e = error
                time.sleep(1)
    
>       raise e
E       AssertionError: assert '1.0' == '2.0'
E         - 1.0
E         + 2.0

test/unit/tools/test_toolbox.py:248: AssertionError
----------------------------- Captured stderr call -----------------------------
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:Created new ToolCache
DEBUG:galaxy.tools.toolbox.watcher:Starting ToolConfWatcher thread
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
INFO:galaxy.tools.toolbox.base:Parsing the tool configuration /tmp/tmplpgMM0/tool_conf.xml
DEBUG:galaxy.tools.cache:Caching tool
DEBUG:galaxy.tools.cache:Done caching tool
DEBUG:galaxy.tools.toolbox.base:Loaded tool id: test_tool, version: 1.0 into tool panel..
DEBUG:galaxy.tools.toolbox.base:Reading tools from config files finished (1.502 ms)
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Removed the following tools from cache: ['test_tool']
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
INFO:galaxy.tools.toolbox.base:Parsing the tool configuration /tmp/tmplpgMM0/tool_conf.xml
ERROR:galaxy.tools.toolbox.base:Tool '/tmp/tmplpgMM0/simple_tool.xml' is not valid:
Traceback (most recent call last):
  File "/home/circleci/repo/lib/galaxy/tools/toolbox/base.py", line 754, in load_tool
    tool = self.create_tool(config_file=config_file, tool_shed_repository=tool_shed_repository, guid=guid, **kwds)
  File "/home/circleci/repo/lib/galaxy/tools/__init__.py", line 272, in create_tool
    raise e
ParseError: syntax error: line 1, column 0
DEBUG:galaxy.tools.cache:Caching tool
DEBUG:galaxy.tools.cache:Done caching tool
DEBUG:galaxy.tools.toolbox.base:Reading tools from config files finished (0.741 ms)
DEBUG:galaxy.tools.toolbox.base:Loaded tool id: test_tool, version: 1.0 into tool panel..
DEBUG:galaxy.tools.toolbox.base:Loading tool panel finished (0.303 ms)
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Removed the following tools from cache: ['test_tool']
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
INFO:galaxy.tools.toolbox.base:Parsing the tool configuration /tmp/tmplpgMM0/tool_conf.xml
DEBUG:galaxy.tools.cache:Caching tool
DEBUG:galaxy.tools.cache:Done caching tool
DEBUG:galaxy.tools.toolbox.base:Reading tools from config files finished (1.566 ms)
DEBUG:galaxy.tools.toolbox.base:Loaded tool id: test_tool, version: 2.0 into tool panel..
DEBUG:galaxy.tools.toolbox.base:Loading tool panel finished (0.474 ms)
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
DEBUG:galaxy.tools.cache:cleaning up
DEBUG:galaxy.tools.cache:Cleanup done
```
So I wondering if this is maybe a scope issue with the assertion function.

mvdbeek committed:
Store toolbox in test_case._toolbox

Apparently accessing double-underscore variables in threads leads
to name-mangling issues, so effectively updating test_case.__toolbox
wouldn't be reflected outside of the watcher thread.
We previously used self.app.toolbox which circumenvented
this issue.
It is possible that `check_tool_errors` / `check_no_tool_errors`
were accessing the old toolbox, unless the watching thread happened
to fire between function definition and assert, and that's probably
why increasing the number of trials had no effect.

mvdbeek committed:
Tear down started threads

mvdbeek commented:
OK, I think this is working now, I suspect https://github.com/galaxyproject/galaxy/pull/7446/commits/c3022c1c055d1ecba4213bdfb02f00a52194d928 fixed it, but the other commits surely make this more robust in general.

nsoranzo commented:
@mvdbeek Fantastic, I'd have never find out what was the issue, thanks a bunch for looking into this!

Maybe we can drop the fsync now?

nsoranzo commented:
Also you may want to rebase to get rid of the circleci commit from the PR that was just merged.

mvdbeek commented:
Sure, let's try that.

nsoranzo commented:
@mvdbeek I've rebased again merging and cleaning the commits, since we have gone back and forth on some changes. Hope that's OK (and more readable in future).

mvdbeek commented:
Sure, np.
